Turn `#[global_allocator]` into a regular attribute macro

petrochenkov opened the pull request:
It was a 99% macro with exception of some diagnostic details.

As a result of the change, `#[global_allocator]` now works in nested modules and even in nameless blocks.

Fixes https://github.com/rust-lang/rust/issues/44113
Fixes https://github.com/rust-lang/rust/issues/58072

petrochenkov committed:
Merge `rustc_allocator` into `libsyntax_ext`

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

petrochenkov committed:
syntax_ext: Turn `#[global_allocator]` into a regular attribute macro

petrochenkov committed:
syntax_ext: Reuse built-in attribute template checking for macro attributes

petrochenkov committed:
syntax_ext: Improve and simplify code generated by `#[global_allocator]`

Instead of
```
mod allocator_abi { /* methods */ }
```
we now generate
```
const _: () = { /* methods */ }
```
and use `std_path` for paths referring to standard library entities.

This way we no longer need to generate `use` and `extern crate` imports, and `#[global_allocator]` starts working inside unnamed blocks.

petrochenkov commented:
r? @eddyb

eddyb commented:
r=me but I'd also check with @alexcrichton

petrochenkov commented:
r? @alexcrichton

alexcrichton commented on src/libsyntax_ext/lib.rs:
```diff
@@ -150,6 +151,12 @@ pub fn register_builtins(resolver: &mut dyn syntax::ext::base::Resolver,
+    register(sym::global_allocator, SyntaxExtension {
+        allow_internal_unstable: Some([sym::rustc_attrs][..].into()),
+        ..SyntaxExtension::default(
+            SyntaxExtensionKind::LegacyAttr(Box::new(global_allocator::expand)), edition
```
I think in theory there's no need for this to be a legacy extension, but is this done for other reasons? I could imagine, for example:

* A bit faster since everything is already parsed
* Guaranteed good error messages since spans are always right
* Other builtin attributes still seem to be using legacy attributes

Just wanted to but sure this was intentional!

petrochenkov replied:
Well, it's legacy because everything built-in is still on legacy (aka AST-based).
We don't have good quasi-quoting that can be used internally to conveniently generate token streams, there are some unresolved issues with error recovery, but the long term plan is to migrate to tokens, yes.

But right now I simply took the existing code from `rustc_allocator` having signature `fn(ast::Item) -> Vec<ast::Item>` and wrapped it into the corresponding AST-based `SyntaxExtensionKind`.

alexcrichton replied:
Ok that's what I figured, thanks!

alexcrichton commented on src/test/ui/feature-gate/issue-43106-gating-of-test.stderr:
```diff
@@ -1,7 +1,13 @@
+error: malformed `test` attribute input
+  --> $DIR/issue-43106-gating-of-test.rs:10:1
+   |
+LL | #![test                    = "4200"]
```
Did we previously silently allow this? If so we may want to do this in a separate PR since the age of `#[test]` means that this could be likely for causing accidental regressions.

petrochenkov replied:
Yes, this was previously allowed.
I don't want to wait for a separate crater run for this, since I don't expect *this* kind of misuse to be common after seeing the statistics from https://github.com/rust-lang/rust/pull/57321.

I think I'll just demote this error to [this lint](https://github.com/rust-lang/rust/issues/57571) for `#[test]` and `#[bench]`.

alexcrichton commented on src/tools/tidy/src/features.rs:
```diff
@@ -344,7 +344,7 @@ fn get_and_check_lib_features(base_src_path: &Path,
                     let mut check_features = |f: &Feature, list: &Features, display: &str| {
                         if let Some(ref s) = list.get(name) {
-                            if f.tracking_issue != s.tracking_issue {
+                            if f.tracking_issue != s.tracking_issue && f.level != Status::Stable {
```
To confirm, was this change intended to be included here?

petrochenkov replied:
Yes, the libcore change doesn't pass tidy without it.
Tidy expected a tracking issue in a `#[stable(...)]` attribute to coincide with something, but the problem is that `#[stable(...)]` doesn't support issue numbers - `#[stable(..., issue = "...")]` is not a valid attribute.

alexcrichton replied:
👍

alexcrichton reviewed (commented):
Thanks @petrochenkov!

Definitely agree this should have been a macro from the start!

This makes me wonder if we can make `#[proc_macro]` and friends macros...

petrochenkov commented:
>This makes me wonder if we can make `#[proc_macro]` and friends macros...

Right now the proc macro harness generated for `--crate-type proc-macro` looks like this:
```rust
#[doc(hidden)]
pub mod decls {
    extern crate proc_macro;
    #[rustc_proc_macro_decls]
    pub static _DECLS: &[proc_macro::bridge::client::ProcMacro] =
        &[proc_macro::bridge::client::ProcMacro::custom_derive("MyDerive",
                                                               &[],
                                                               crate::my_derive),
          proc_macro::bridge::client::ProcMacro::attr("my_attr",
                                                      crate::my_attr),
          proc_macro::bridge::client::ProcMacro::bang("my_macro",
                                                      crate::my_macro)];
}
```
So, it cannot be generated by individual `#[proc_macro*]` attributes without a rework of the proc macro crate ABI.

alexcrichton commented:
Ok r=me if you're confident `#[test]` won't cause much breakage, it's easy to back out anyway. If you'd like to change it to wire up to a lint r=me also on that.

petrochenkov committed:
Demote template check error to a lint for `#[test]` and `#[bench]`

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit d2726fbbbe9c0ce97f4c3458185b0e7846c1f0bb has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d2726fbbbe9c0ce97f4c3458185b0e7846c1f0bb -->
<!-- homu: {"type":"Approved","sha":"d2726fbbbe9c0ce97f4c3458185b0e7846c1f0bb","approver":"alexcrichton"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #62705) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov committed:
Fix rebase

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 54041c60bdc889ccf29dd57efde413515950bb2e has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 54041c60bdc889ccf29dd57efde413515950bb2e -->
<!-- homu: {"type":"Approved","sha":"54041c60bdc889ccf29dd57efde413515950bb2e","approver":"alexcrichton"} -->

eddyb commented:
@petrochenkov That's interesting, so proc macros and tests could use a common "gather" infrastructure? Then again, we could just store all the symbol names, and other information like attributes, entirely in crate metadata, and have `rustc_metadata` load it all based on that?

The only catch is that the symbol name can't be of the function itself, but of a wrapper around it (which could live in a hidden static next to the function).

petrochenkov commented:
@eddyb 
>so proc macros and tests could use a common "gather" infrastructure? Then again, we could just store all the symbol names, and other information like attributes, entirely in crate metadata, and have rustc_metadata load it all based on that?

These two are similar, after the recent refactorings they are the only passes that
- cannot be represented as individual macros
- must run after the usual expansion is complete (this causes most of hacks and issues)

I didn't thought about the proc macro infra much, but regarding the test harness, it has to jump through so many hoops, and has so many silly restrictions, that it certainly comes to mind that it operates at the wrong time on the wrong data.

Perhaps it shouldn't construct all those imports on top of the already converged system and attempt to resolve them, but rather work e.g. at HIR lowering time and directly build some structures specialized for testing from available AST and `Res`olutions.
Yes, that would move some knowledge about the `--test` mode to HIR, but it could make test harness more reliable, and working in more scenarios, and hopefully less hacky.

eddyb commented:
@petrochenkov So @Manishearth and someone else (I forget who now, sorry) were discussing the test harness stuff and the solution that seemed most reasonable was to add one `const` item per test (with whatever type `libtest` wants), that the compiler than gathers into a const slice at the top of the crate (even if the individual constants are inaccessible to imports).
That mechanism is pretty general and could work with whatever types a harness needed.

bors commented:
:umbrella: The latest upstream changes (presumably #62902) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit a0c2c640d54fa1622c2fea4accae1025bf109c47 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton a0c2c640d54fa1622c2fea4accae1025bf109c47 -->
<!-- homu: {"type":"Approved","sha":"a0c2c640d54fa1622c2fea4accae1025bf109c47","approver":"alexcrichton"} -->
